strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java

```diff
@@ -39,29 +47,99 @@ public void postConstruct()
         entityServiceRegistry.register(this.getEntityClass(), this);
     }
 
-    // TODO: recursive identification + collections
-    protected void cascadeEntityIdentification(T entity)
+    protected <S extends T> S cascadeEntitySave(T entity)
     {
         identifyEntity(entity);
-        ReflectionUtils.doWithFields(entity.getClass(), (field)->{
+        
+        ReflectionUtils.doWithFields(entity.getClass(), (field) -> {
             ReflectionUtils.makeAccessible(field);
-            Class<? extends GenericEntity> t = (Class<? extends GenericEntity>) field.getType();
-            if (!GenericEntity.class.isAssignableFrom(t))
+
+            Set<CascadeType> cascadeTypeSet = Arrays.stream(field.getAnnotations())
+                                                    .map(a -> exposeCascadeType(a))
+                                                    .reduce((c1,
+                                                             c2) -> {
+                                                        c1.addAll(c2);
+                                                        return c1;
+                                                    })
+                                                    .orElse(Collections.emptySet());
+            
+            if (!cascadeTypeSet.stream().anyMatch(c -> CascadeType.ALL.equals(c) || CascadeType.MERGE.equals(c)
+                    || CascadeType.PERSIST.equals(c)))
             {
                 return;
             }
             
-            GenericEntity subEntity = (GenericEntity) ReflectionUtils.getField(field, entity);
-            if (subEntity == null)
+            Class<?> fieldType = field.getType();
+            Object fieldValue = ReflectionUtils.getField(field, entity);
+
+            if (fieldValue == null)
             {
                 return;
             }
 
-            CommonCrudService<GenericEntity> entityService = (CommonCrudService<GenericEntity>) entityServiceRegistry.getEntityService(subEntity.getClass());
-            entityService.identifyEntity(subEntity);
+            if (Collection.class.isAssignableFrom(fieldType))
```

Comment:
Will we support `Map`-s ? For example, `org.carlspring.strongbox.configuration.Configuration#storages`

Reply:
`org.carlspring.strongbox.configuration.Configuration#storages` is `@Embeded`.